editor/plugins/curve_editor_plugin.cpp

```diff
@@ -33,276 +33,354 @@
+			point_radius = Math::round(BASE_POINT_RADIUS * get_theme_default_base_scale() * gizmo_scale);
+			hover_radius = Math::round(BASE_HOVER_RADIUS * get_theme_default_base_scale() * gizmo_scale);
+			tangent_radius = Math::round(BASE_TANGENT_RADIUS * get_theme_default_base_scale() * gizmo_scale);
+			tangent_hover_radius = Math::round(BASE_TANGENT_HOVER_RADIUS * get_theme_default_base_scale() * gizmo_scale);
+			tangent_length = Math::round(BASE_TANGENT_LENGTH * get_theme_default_base_scale());
+		} break;
 		case NOTIFICATION_DRAW: {
 			_draw();
 		} break;
-		case EditorSettings::NOTIFICATION_EDITOR_SETTINGS_CHANGED: {
-			if (EditorSettings::get_singleton()->check_changed_settings_in_group("interface/touchscreen/scale_gizmo_handles")) {
-				_gizmo_handle_scale = EDITOR_GET("interface/touchscreen/scale_gizmo_handles");
+		case NOTIFICATION_VISIBILITY_CHANGED: {
+			if (!is_visible()) {
+				grabbing = GRAB_NONE;
 			}
 		} break;
 	}
 }
 
-void CurveEditor::gui_input(const Ref<InputEvent> &p_event) {
-	Ref<InputEventMouseButton> mb_ref = p_event;
-	if (mb_ref.is_valid()) {
-		const InputEventMouseButton &mb = **mb_ref;
+void CurveEdit::gui_input(const Ref<InputEvent> &p_event) {
+	ERR_FAIL_COND(p_event.is_null());
+
+	Ref<InputEventKey> k = p_event;
+	if (k.is_valid()) {
+		// Deleting points or making tangents linear.
+		if (k->is_pressed() && k->get_keycode() == Key::KEY_DELETE) {
+			if (selected_tangent_index != TANGENT_NONE) {
+				toggle_linear(selected_index, selected_tangent_index);
+			} else if (selected_index != -1) {
+				if (grabbing == GRAB_ADD) {
+					curve->remove_point(selected_index); // Point is temporary, so remove directly from curve.
+					set_selected_index(-1);
+				} else {
+					remove_point(selected_index);
+				}
+				grabbing = GRAB_NONE;
+				hovered_index = -1;
+				hovered_tangent_index = TANGENT_NONE;
+			}
+			accept_event();
+		}
+
+		if (k->get_keycode() == Key::SHIFT || k->get_keycode() == Key::ALT) {
+			queue_redraw();
+		}
+	}
 
-		if (mb.is_pressed() && !_dragging) {
-			Vector2 mpos = mb.get_position();
+	Ref<InputEventMouseButton> mb = p_event;
+	if (mb.is_valid() && mb->is_pressed()) {
+		Vector2 mpos = mb->get_position();
+
+		if (mb->get_button_index() == MouseButton::RIGHT || mb->get_button_index() == MouseButton::MIDDLE) {
+			if (mb->get_button_index() == MouseButton::RIGHT && grabbing == GRAB_MOVE) {
+				// Move a point to its old position.
+				curve->set_point_value(selected_index, initial_grab_pos.y);
+				curve->set_point_offset(selected_index, initial_grab_pos.x);
+				set_selected_index(initial_grab_index);
+				hovered_index = -1;
+				grabbing = GRAB_NONE;
+			} else {
+				// Remove a point or make a tangent linear.
+				selected_tangent_index = get_tangent_at(mpos);
+				if (selected_tangent_index != TANGENT_NONE) {
+					toggle_linear(selected_index, selected_tangent_index);
+				} else {
+					int point_to_remove = get_point_at(mpos);
```

Comment:
Fair enough, hehe :)

Reply:
Ah right. Because I couldn't ensure the hovering logic is always correct and updated. So it's better to do this and keep `hovered_index` for visual stuff.

Reply:
Ah, makes sense - keep the visuals separated from the actual logic so you don't have to be always super on top of it everywhere in code! :)